Firestore/core/src/firebase/firestore/model/document_key.h

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright 2018 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_MODEL_DOCUMENT_KEY_H_
+#define FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_MODEL_DOCUMENT_KEY_H_
+
+#include <initializer_list>
+#include <string>
+
+#include "Firestore/core/src/firebase/firestore/model/resource_path.h"
+#include "absl/strings/string_view.h"
+
+namespace firebase {
+namespace firestore {
+namespace model {
+
+class DocumentKey {
+ public:
+  DocumentKey() = default;
+  /** Creates and returns a new document key with a copy of the given path. */
+  explicit DocumentKey(const ResourcePath& path);
+  /**
+   * Creates and returns a new document key, taking ownership of the given
+   * path.
+   */
+  explicit DocumentKey(ResourcePath&& path);
+
+  /**
+   * Creates and returns a new document key using '/' to split the string into
+   * segments.
+   */
+  static DocumentKey FromPathString(absl::string_view path) {
+    return DocumentKey{ResourcePath::FromString(path)};
+  }
+  /**
+   * Creates and returns a new document key with the given segments.
+   */
+  static DocumentKey FromSegments(std::initializer_list<std::string> list) {
+    return DocumentKey{ResourcePath{list}};
+  }
+
+  /**
+   * Returns a shared instance of an empty document key.
+   */
+  static const DocumentKey& Empty();
+
+  static bool IsDocumentKey(const ResourcePath& path) {
+    return path.size() % 2 == 0;
+  }
+
+  const ResourcePath& path() const {
+    return path_;
+  }
+
+ private:
+  ResourcePath path_;
```

Comment:
DocumentKeys are frequently used as keys in maps and extracted from the documents they name. Given that these are not that cheap to create (they contain a vector of strings) I wonder if we should make DocumentKey a pointer-hiding value type that contains a `std::shared_ptr<ResourcePath>` instead of just a `ResourcePath`.

This would make DocumentKeys substantially cheaper to pass around and would avoid is making maps like this:

std::unordered_map<std::shared_ptr<Document_key>, std::shared_ptr<Document>>

Reply:
What's the bigger concern, copying keys around or looking them up in a map? For lookup, wouldn't shared pointer make it slower because of an additional indirection (or not affect the time, in best case)?

If lookup is the bigger concern, then perhaps we should cache the hash (since DocumentKey is *mostly* immutable, it shouldn't change often)? If copying is the bigger concern, then shared pointer sounds good. Let me look more at the usages and get back to you.

Reply:
Sorry, the concern I was trying to raise was copying.

If you look through the LocalStore we do a whole bunch of operations on documents track the affected keys. Similarly we compute views using sets of affected keys. All of these come from the underlying documents.

The `FSTDocumentKey` we're replacing, as an Objective-C object, is always heap allocated and always effectively tracked through shared pointers (that's what a strong reference under ARC effectively is). This means that making `DocumentKey` contain a `shared_ptr` we'll end up essentially no worse off that we were before.

Reply:
This makes shared pointer sound very reasonable. Is hashing of no concern at all (the two aren't mutually exclusive). Also, is FSTDocumentKey passed between threads?

Reply:
Just to throw an idea around, shared_ptr optimization could be moved to BasePath as well. That will make it affect the public FieldPath, too, for better or worse.

Reply:
FSTDocumentKey (by virtue of being part of FSTDocument) is passed between threads when we deliver snapshots to users. This is part of why I'm insistent that these structures be (mostly) immutable.

Unlike keys I don’t think field paths get shared like this so I’m not sure it’s worth it. Keys are an obvious win case.

We can always refactor later if we find field paths are getting copied around.

Reply:
Done.

Note that it means that DocumentKey may now be in an invalid moved-from state. To make the behavior match the previous iteration better, I did:

```
const ResourcePath& DocumentKey::path() const {
  return path_ ? *path_ : Empty().path();
}
```

which effectively guarantees that moved-from state is always valid.